<a href="https://colab.research.google.com/github/boiBASH/Sterling-Bank-Data-Science-Assessment./blob/main/Data_Preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
#Import Necessary Libaries
import pandas as pd
from datetime import datetime
from sklearn.preprocessing import LabelEncoder
from IPython.display import display

In [2]:
#Loading dataset
df = pd.read_excel('/content/drive/MyDrive/Sterling Bank Assessment/data_science_assessment_data (1).xlsx')
df

loan_id       sector                           industry_x  \
0     AA1818056YFR  AGRICULTURE  Agro Services                         
1     AA1818056YFR  AGRICULTURE  Agro Services                         
2     AA17088XW27N  AGRICULTURE  Agro Services                         
3     AA17094XPZ3W  AGRICULTURE  Food Crops ,grains                    
4     AA17094XPZ3W  AGRICULTURE  Food Crops ,grains                    
...            ...          ...                                  ...   
1575  AA21326S23X3      GENERAL  Food Crops ,grains                    
1576  AA231813L1ZQ      GENERAL  Food Crops ,grains                    
1577  AA231813L1ZQ      GENERAL  Food Crops ,grains                    
1578  AA1613000NG1  AGRICULTURE  Fishing                               
1579  AA163219BYYX  AGRICULTURE  Agro Services                         

            PRODUCT_GROUP_NAME                     LINE_DESC FACILITY_TYPE  \
0        Corporate Loans Group        BILLS DUE - TERM LOANS     TERM LOAN   
1        Corporate Loans Group        BILLS DUE - TERM LOANS     TERM LOAN   
2     Agricultural Loans Group       BILLS DUE - AGRIC LOANS    OTHER LOAN   
3     Agricultural Loans Group       BILLS DUE - AGRIC LOANS    OTHER LOAN   
4     Agricultural Loans Group       BILLS DUE - AGRIC LOANS    OTHER LOAN   
...                        ...                           ...           ...   
1575  Agricultural Loans Group       BILLS DUE - AGRIC LOANS    OTHER LOAN   
1576  Agricultural Loans Group    MSME AGRIC FINANCE - WOMEN    OTHER LOAN   
1577  Agricultural Loans Group    MSME AGRIC FINANCE - WOMEN    OTHER LOAN   
1578    Current Accounts Group  STERLING SUPA AC- INDIVIDUAL     OVERDRAFT   
1579  Agricultural Loans Group       BILLS DUE - AGRIC LOANS    OTHER LOAN   

      Contract_Amount      CURR_BAL  EQUIV_BALANCE Contract_eff_date  ...  \
0        6.252801e+05  1.217761e+05   1.217761e+05        2018-06-29  ...   
1        6.252801e+05  1.217761e+05   1.217761e+05        2018-06-29  ...   
2        1.801259e+04  4.200000e+00   4.200000e+00        2017-03-29  ...   
3        5.176228e+06  1.975372e+06   1.975372e+06        2017-04-04  ...   
4        5.176228e+06  1.975372e+06   1.975372e+06        2017-04-04  ...   
...               ...           ...            ...               ...  ...   
1575     2.452215e+06  2.452215e+06   2.452215e+06        2021-11-22  ...   
1576     2.786393e+06  2.786393e+06   2.786393e+06        2023-06-30  ...   
1577     2.786393e+06  2.786393e+06   2.786393e+06        2023-06-30  ...   
1578     1.000000e+07  9.745611e+06   9.745611e+06        2016-05-09  ...   
1579     5.222775e+09  2.637027e+08   2.637027e+08        2016-11-16  ...   

     running_loans_count running_loans_amount previous_loan_default_count  \
0                      0                 0.00                           1   
1                      0                 0.00                           1   
2                      0                 0.00                           2   
3                      0                 0.00                           1   
4                      0                 0.00                           1   
...                  ...                  ...                         ...   
1575                   1           2786392.55                           1   
1576                   1           2786392.55                           1   
1577                   1           2786392.55                           1   
1578                   0                 0.00                           1   
1579                   0                 0.00                           1   

     date_of_birth marital_status  employment_status  customer_inception_date  \
0       1991-11-29            NaN           EMPLOYED               2019-11-01   
1       1991-11-29            NaN           EMPLOYED               2015-12-17   
2              NaT            NaN                NaN               2016-08-24   
3              NaT            NaN  

In [3]:
# Dataset overview
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1580 entries, 0 to 1579
Data columns (total 36 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   loan_id                      1580 non-null   object        
 1   sector                       1580 non-null   object        
 2   industry_x                   1580 non-null   object        
 3   PRODUCT_GROUP_NAME           1580 non-null   object        
 4   LINE_DESC                    1580 non-null   object        
 5   FACILITY_TYPE                1580 non-null   object        
 6   Contract_Amount              1580 non-null   float64       
 7   CURR_BAL                     1574 non-null   float64       
 8   EQUIV_BALANCE                1574 non-null   float64       
 9   Contract_eff_date            1580 non-null   datetime64[ns]
 10  CONTRACT_MAT_DATE            1580 non-null   datetime64[ns]
 11  report_date                  1580 non-null 

# **Data Summary**
The dataset contains 1,580 rows and 36 columns, covering various aspects of loan applicants and loan characteristics.

In [4]:
df.describe()

Contract_Amount      CURR_BAL  EQUIV_BALANCE  \
count     1.580000e+03  1.574000e+03   1.574000e+03   
mean      3.461629e+08  2.380443e+08   2.380443e+08   
min       1.022630e+03  0.000000e+00   0.000000e+00   
25%       1.450000e+06  3.070457e+05   3.070457e+05   
50%       5.000000e+06  2.100000e+06   2.100000e+06   
75%       6.736339e+07  2.000000e+07   2.000000e+07   
max       2.017653e+10  1.927179e+10   1.927179e+10   
std       1.814535e+09  1.540650e+09   1.540650e+09   

                   Contract_eff_date              CONTRACT_MAT_DATE  \
count                           1580                           1580   
mean   2022-01-16 21:29:37.215189760  2023-04-27 11:36:18.227848192   
min              2016-04-14 00:00:00            2016-06-13 00:00:00   
25%              2021-10-20 00:00:00            2022-12-02 00:00:00   
50%              2022-09-05 00:00:00            2023-06-04 00:00:00   
75%              2023-02-23 06:00:00            2023-11-03 00:00:00   
max              2023-08-23 00:00:00            2037-05-31 00:00:00   
std                              NaN                            NaN   

                         report_date     PayinAccount_Last_LOD_Date  \
count                           1580                           1505   
mean   2023-02-20 08:52:15.189873152  2023-02-15 01:11:45.647840512   
min              2018-04-30 00:00:00            2017-06-22 00:00:00   
25%              2023-01-16 00:00:00            2022-12-23 00:00:00   
50%              2023-06-30 00:00:00            2023-06-01 00:00:00   
75%              2023-08-23 00:00:00            2023-08-08 00:00:00   
max              2023-08-23 00:00:00            2023-08-25 00:00:00   
std                              NaN                            NaN   

       MONTH_LOAN_DURATION  YEAR_LOAN_DURATION  DAYS_RUNNING_SINCE  ...  \
count          1580.000000         1580.000000         1580.000000  ...   
mean             15.175949            1.248734          399.474051  ...   
min               0.000000            0.000000            0.000000  ...   
25%               4.000000            0.000000           94.000000  ...   
50%              11.000000            1.000000          209.000000  ...   
75%              17.000000            1.000000          453.500000  ...   
max             240.000000           20.000000         2675.000000  ...   
std              21.362445            1.802687          507.856083  ...   

       DAYS_TO_MATURITY  Default_status  previous_loans_count  \
count       1580.000000     1580.000000           1580.000000   
mean          66.113924        0.344937              3.983544   
min        -2206.000000        0.000000              0.000000   
25%          -12.000000        0.000000              1.000000   
50%            5.000000        0.000000              3.000000   
75%          117.250000        1.000000              5.000000   
max         5030.000000        1.000000             34.000000   
std          486.567275        0.475498              4.187316   

       previous_loans_amount  running_loans_count  running_loans_amount  \
count           1.580000e+03          1580.000000          1.580000e+03   
mean            1.802177e+09             1.229114          7.561980e+08   
min             0.000000e+00             0.000000          0.000000e+00   
25%             1.800000e+06             0.000000          0.000000e+00   
50%             8.723279e+06             1.000000          2.000000e+06   
75%             1.600000e+08             2.000000          2.000000e+07   
max             8.705381e+10             7.000000          4.035307e+10   
std             1.067693e+10             1.534707          3.750349e+09   

       previous_loan_default_count                  date_of_birth  \
count                  1580.000000                            946   
mean                      1.491772  1991-02-27 18:09:53.657505280   
min                       0.000000            1952-01-29 00:00:00   
25%              

In [5]:
#Checking for missing values
missing = df.isnull().sum().to_frame('missing_count')
missing

missing_count
loan_id                                  0
sector                                   0
industry_x                               0
PRODUCT_GROUP_NAME                       0
LINE_DESC                                0
FACILITY_TYPE                            0
Contract_Amount                          0
CURR_BAL                                 6
EQUIV_BALANCE                            6
Contract_eff_date                        0
CONTRACT_MAT_DATE                        0
report_date                              0
PayinAccount_Last_LOD_Date              75
ARR_STATUS                               0
INTEREST_RATE                           57
MONTH_LOAN_DURATION                      0
YEAR_LOAN_DURATION                       0
DAYS_RUNNING_SINCE                       0
DAY_SINCEPAYMENT                        75
DAYS_TO_MATURITY                         0
DIRECTORATENAME                        193
Is_Active_loans                          0
Default_status                           0
Default_status_kind                      0
previous_loans_count                     0
previous_loans_amount                    0
running_loans_count                      0
running_loans_amount                     0
previous_loan_default_count              0
date_of_birth                          634
marital_status                        1516
employment_status                      838
customer_inception_date                  0
industry_y                               0
industry_desc                            0
sector_desc                              0

In [6]:
# Missing-value summary
missing_summary = df.isnull().sum().sort_values(ascending=False)
missing_percent = (missing_summary / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing Count': missing_summary,
    'Missing %': missing_percent
})
missing_df = missing_df[missing_df['Missing Count'] > 0]
missing_df

Missing Count  Missing %
marital_status                       1516  95.949367
employment_status                     838  53.037975
date_of_birth                         634  40.126582
DIRECTORATENAME                       193  12.215190
DAY_SINCEPAYMENT                       75   4.746835
PayinAccount_Last_LOD_Date             75   4.746835
INTEREST_RATE                          57   3.607595
CURR_BAL                                6   0.379747
EQUIV_BALANCE                           6   0.379747

# **Data Cleaning Highlighted In Steps**

In [7]:
# Step 1: Drop 'marital_status' due to very high missing rate
df.drop(columns=['marital_status'], inplace=True)

# Step 2: Replace missing 'employment_status' with 'Unknown'
df['employment_status'].fillna('Unknown', inplace=True)

# Step 3: Calculate age from 'date_of_birth' and drop the original column
today = pd.to_datetime("today")
df['age'] = df['date_of_birth'].apply(lambda x: (today - x).days // 365 if pd.notnull(x) else None)
df.drop(columns=['date_of_birth'], inplace=True)

# Step 4: Impute 'DAY_SINCEPAYMENT' with median
median_payment_days = df['DAY_SINCEPAYMENT'].median()
df['DAY_SINCEPAYMENT'].fillna(median_payment_days, inplace=True)

# Step 5: Fill missing 'DIRECTORATENAME' with 'Unknown'
df['DIRECTORATENAME'].fillna('Unknown', inplace=True)

# Step 6: Drop identifier and redundant columns
columns_to_drop = [
    'loan_id',          # identifier
    'industry_x',       # redundant with industry_desc
    'industry_y',       # numeric code for industry
    'sector_desc',      # duplicate of 'sector'
    'industry_desc'     # keep only if highly detailed vs 'LINE_DESC'
]
df.drop(columns=columns_to_drop, inplace=True)

# Step 7: Feature Engineering from dates
today = pd.to_datetime("today")
df['loan_age_days'] = (today - df['Contract_eff_date']).dt.days
df['customer_tenure_days'] = (today - df['customer_inception_date']).dt.days

# Step 8: Drop raw date columns used in feature engineering
df.drop(columns=['Contract_eff_date', 'customer_inception_date'], inplace=True)

# Preview the cleaned dataset before EDA
df_ready_preview = df.head()
df_ready_preview

/tmp/ipython-input-2786703129.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['employment_status'].fillna('Unknown', inplace=True)
/tmp/ipython-input-2786703129.py:14: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)'

sector        PRODUCT_GROUP_NAME                LINE_DESC  \
0  AGRICULTURE     Corporate Loans Group   BILLS DUE - TERM LOANS   
1  AGRICULTURE     Corporate Loans Group   BILLS DUE - TERM LOANS   
2  AGRICULTURE  Agricultural Loans Group  BILLS DUE - AGRIC LOANS   
3  AGRICULTURE  Agricultural Loans Group  BILLS DUE - AGRIC LOANS   
4  AGRICULTURE  Agricultural Loans Group  BILLS DUE - AGRIC LOANS   

  FACILITY_TYPE  Contract_Amount    CURR_BAL  EQUIV_BALANCE CONTRACT_MAT_DATE  \
0     TERM LOAN        625280.13   121776.05      121776.05        2020-06-24   
1     TERM LOAN        625280.13   121776.05      121776.05        2020-06-24   
2    OTHER LOAN         18012.59        4.20           4.20        2017-08-26   
3    OTHER LOAN       5176227.76  1975372.41     1975372.41        2018-02-28   
4    OTHER LOAN       5176227.76  1975372.41     1975372.41        2018-02-28   

  report_date PayinAccount_Last_LOD_Date  ... Default_status_kind  \
0  2023-05-05                 2020-02-07  ...                LOST   
1  2023-05-05                 2020-02-07  ...                LOST   
2  2021-05-31                 2019-06-28  ...                LOST   
3  2021-05-31                 2017-07-19  ...                LOST   
4  2021-05-31                 2017-07-19  ...                LOST   

  previous_loans_count  previous_loans_amount  running_loans_count  \
0                    4             1165280.13                    0   
1                    4             1165280.13                    0   
2                    2               36691.40                    0   
3                    1             5176227.76                    0   
4                    1             5176227.76                    0   

   running_loans_amount  previous_loan_default_count  employment_status   age  \
0                   0.0                            1           EMPLOYED  33.0   
1                   0.0                            1           EMPLOYED  33.0   
2                   0.0                            2            Unknown   NaN   
3                   0.0                            1            Unknown   NaN   
4                   0.0                            1            Unknown   NaN   

  loan_age_days  customer_tenure_days  
0          2591                  2101  
1          2591                  3516  
2          3048                  3265  
3          3042                  3327  
4          3042                  3043  

[5 rows x 30 columns]

In [8]:
#Save Clean Data for Exploratory Data Analysis (EDA)
cleaned_path = "/content/drive/MyDrive/Sterling Bank Assessment/cleaned_loan_data.xlsx"
df.to_excel(cleaned_path, index=False)

cleaned_path

'/content/drive/MyDrive/Sterling Bank Assessment/cleaned_loan_data.xlsx'